File: python/data-structures/vector2d.py
import math

# Vector2D(number, number) -> Vector2D
# Constructs a two-dimensional vector object.
# v = Vector2D(3, 4)
# v.x -> 3
# v.y -> 4
# v.point -> (3, 4)
# v.magnitude -> 5
# v.unit_vector -> (0.6, 0.8)

class Vector2D:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.point = (x, y)
        self.magnitude = math.hypot(x, y)

    def unit_vector(self):
        x, y = self.point
        mag = self.magnitude
        return (x / mag, y / mag)
    
# scale(Vector2D, number) -> Vector2D
# Returns a scaled two-dimensional vector that is the product of a vector and a number.
# v = scale(Vector2D(3, 4), 2) ->  v.point -> (6, 8)

def scale(v, factor):
    x, y = v.point
    return Vector2D((x * factor), (y * factor))

# add(Vector2D, Vector2D) -> Vector2D
# Returns a two-dimensional vector that is the sum of two vectors.
# v = add(Vector2D(1, 2), Vector2D(3, 4)) -> v.point -> (4, 6)

def add(v1, v2):
    x1, y1 = v1.point
    x2, y2 = v2.point
    return Vector2D(x1 + x2, y1 + y2)

# dot_product(Vector2D, Vector2D) -> number
# Returns a number that is the dot product of two, two-dimensional vectors.
# dot_product(Vector2D(1, 2), Vector2D(3, 4)) -> 11

def dot_product(v1, v2):
    x1, y1 = v1.point
    x2, y2 = v2.point
    return (x1 * x2) + (y1 * y2)

# cross_product(Vector2D, Vector2D) -> number
# Returns a number that is the cross product of two, two-dimensional vectors.
# cross_product(Vector2D(1, 2), Vector2D(3, 4)) -> -2

def cross_product(v1, v2):
    x1, y1 = v1.point
    x2, y2 = v2.point
    return (x1 * y2) - (y1 * x2)

# compare(function) -> function -> boolean
# Generates functions for sequentially comparing the magnitudes of a series of two-dimensional vectors.
# eq = compare(lambda x, y: x == y) -> eq(Vector2D(3, 4), Vector2D(3, 4)) -> True

def compare(operation):
    def operator(v1, v2):
        m1 = v1.magnitude
        m2 = v2.magnitude
        return operation(m1, m2)
    return operator

gt = compare(lambda x, y: x > y)
lt = compare(lambda x, y: x < y)
eq = compare(lambda x, y: x == y)

File: golang/definitions/vector2d.go
package definitions

import (
	"fmt"
	"math"
)

// A two-dimensional quantity that has direction and magnitude. Represented by coordinates X and Y.
type Vector2d struct {
	X, Y float64
}
// Ihat, Jhat
// Mutually orthogonal unit vectors, forming the standard basis.
var (
	Ihat = Vector2d{1, 0}
	Jhat = Vector2d{0, 1}
)

// Fulfills the Stringer interface for the fmt package.
func (v Vector2d) String() string {
	x := v.X
	y := v.Y
	return fmt.Sprintf("vec(%.2f, %.2f)", x, y)
}

// Computes the distance of a 2d-vector's point from the origin.
func (v Vector2d) Magnitude() float64 {
	x := v.X
	y := v.Y
	return math.Hypot(x, y)
}

// Scales a 2d-vector by a factor of n.
func Scale(v Vector2d, n float64) Vector2d {
	x := v.X
	y := v.Y
	return Vector2d{x * n, y * n}
}

// Returns the sum of a series of two-dimensional vectors.
func Add(vs ...Vector2d) Vector2d {
	sum := Vector2d{0, 0}
	for _, v := range vs {
		x1 := sum.X
		y1 := sum.Y
		x2 := v.X
		y2 := v.Y
		sum = Vector2d{x1 + x2, y1 + y2}
	}
	return sum
}

// Computes the dot product of two 2d-vectors.
func DotProduct(v1, v2 Vector2d) float64 {
	x1 := v1.X
	y1 := v1.Y
	x2 := v2.X
	y2 := v2.Y
	return x1*x2 + y1*y2
}

// Computes the cross product of two 2d-vectors.
func CrossProduct(v1, v2 Vector2d) float64 {
	x1 := v1.X
	y1 := v1.Y
	x2 := v2.X
	y2 := v2.Y
	return x1*y2 - y1*x2
}

// Generates functions for comparing the magnitudes of two 2d-vectors.
func compare(op func(float64, float64) bool) func(Vector2d, Vector2d) bool {
	return func(v1, v2 Vector2d) bool {
		m1 := v1.Magnitude()
		m2 := v2.Magnitude()
		return op(m1, m2)
	}
}

// 2d-Vector comparison functions
var Gt = compare(func(x, y float64) bool { return x > y })
var Lt = compare(func(x, y float64) bool { return x < y })
var Eq = compare(func(x, y float64) bool { return x == y })
